sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,334 @@
+      for (Row row : resultSet) {
+        TokenRange tokenRange =
+            new TokenRange(
+                row.getLong("partitions_count"),
+                row.getLong("mean_partition_size"),
+                row.getLong("range_start"),
+                row.getLong("range_end"));
+        tokenRanges.add(tokenRange);
+      }
+      // The table may not contain the estimates yet
+      // or have partitions_count and mean_partition_size fields = 0
+      // if the data was just inserted and the amount of data in the table was small.
+      // This is very common situation during tests,
+      // when we insert a few rows and immediately query them.
+      // However, for tiny data sets the lack of size estimates is not a problem at all,
+      // because we don't want to split tiny data anyways.
+      // Therefore, we're not issuing a warning if the result set was empty
+      // or mean_partition_size and partitions_count = 0.
+      return tokenRanges;
+    } catch (InvalidQueryException e) {
+      LOG.warn("Failed to fetch size estimates for {}.{} from system.size_estimates table. "
+          + "The number of created bundles may be inaccurate. Please make sure you use "
+          + "Cassandra 2.1.5 or newer.", keyspace, table, e);
+      return Arrays.asList(
+          new TokenRange(0L, 0L, 0L, 0L));
+    }
+  }
+
+  /**
+   * Compute the percentage of token addressed compared with the whole tokens in the cluster.
+   */
+  @VisibleForTesting
+  protected static BigDecimal getRingFraction(Cluster cluster, List<TokenRange> tokenRanges) {
+    Partitioner partitioner = new Partitioner(cluster);
+    BigInteger addressedTokens = BigInteger.valueOf(0);
+    for (TokenRange tokenRange : tokenRanges) {
+      addressedTokens = addressedTokens.add(
+          distance(cluster, tokenRange.rangeStart, tokenRange.rangeEnd));
+    }
+    return new BigDecimal(addressedTokens).divide(new BigDecimal
+        (partitioner.getTotalTokens()), 2, RoundingMode.HALF_UP);
+  }
+
+  /**
+   * Measure distance between two tokens.
+   */
+  @VisibleForTesting
+  protected static BigInteger distance(Cluster cluster, long token1, long token2) {
+    Partitioner partitioner = new Partitioner(cluster);
+    if (token2 > token1) {
+      return BigInteger.valueOf(token2).subtract(BigInteger.valueOf(token1));
+    } else {
+      return BigInteger.valueOf(token2).subtract(BigInteger.valueOf(token1))
+          .add(partitioner.getTotalTokens());
+    }
+  }
+
+  static class Partitioner {
+    BigInteger minToken;
+    BigInteger maxToken;
+
+    Partitioner(Cluster cluster) {
+      String partitionerName = cluster.getMetadata().getPartitioner();
+      switch (partitionerName) {
+        case "org.apache.cassandra.dht.Murmur3Partitioner":
+          minToken = BigInteger.valueOf(Long.MIN_VALUE);
+          maxToken = BigInteger.valueOf(Long.MAX_VALUE);
+          break;
+        case "org.apache.cassandra.dht.RandomPartitioner":
```

Comment:
Because the `RandomPartitioner` (used in Cassandra 1.x) has different token range then the new `Murmur3Partitioner` (from Cassandra >= 2.5).